### ISSUE TITLE
Browsing history entries in admin fires pre/post save signals potentially leading to nasty side effects

### ISSUE THREAD

**Issue by romanek-adam-b2c2** [etianen/django-reversion#936]
As described in the subject - opening a specific history entry using django admin causes the underlying model pre/post save signals to be fired which may potentially lead to nasty side effects.

In our case we have a `post_save` signal handler attached to one of our models. The signal handler takes an instance of a model, extracts some fields from it and publishes them to Redis. The expectation is that Redis would always keep the latest "view" on a subset of the model instance fields.

However, I realised that when browsing the history of a specific model instance via django admin, historical instances are de-serialised from JSON and temporarily saved inside a database transaction (using savepoints) BUT this also leads to firing normal pre/post save signals. So in our case our signal handler publishes to Redis the history entry which was opened last. It totally breaks our business logic and thus turned out to be a critical bug.

I'm fairly sure this side effect is not obvious to most users of this library but it's not documented anywhere. Am I missing something?

EDIT: when browsing history using django admin the pre/post signals are fired with `raw` param set to `True`. This param can be used to distinguish them from the regular signals fired when saving a model instance - they get `raw=False`.

**Comment by etianen** [comment]
That is correct, and definitely counts as a gotcha for the admin integration. 😢

**Comment by etianen** [comment]
I've added a note to the top of the admin docs about this 🙇 
https://django-reversion.readthedocs.io/en/latest/admin.html

**Comment by romanek-adam-b2c2** [comment]
Thank you @etianen !

**Comment by etianen** [comment]
[That's a good idea](https://github.com/etianen/django-reversion/commit/48972c143c5be22917cb70fff1e894d584ce4fee) 🙇

**Comment by romanek-adam-b2c2** [comment]
Hey @etianen this issue has hit us again :(

This is because the `raw` param is only present in pre/post -save signals and it's not preset in other signals like `m2m_changed`. Unfortunately, our handler for `m2m_changed` was causing side-effects without using `transaction.on_commit()` so effectively these side-effects were run as part of browsing the history entries in admin.

For now, as a quick workaround we implemented the following change:
```python
from contextlib import nullcontext

import factory
from django.db.models.signals import (
    m2m_changed,
    post_delete,
    post_save,
    pre_delete,
    pre_save,
)
from django.http import HttpRequest, HttpResponse
from reversion_compare.admin import CompareVersionAdmin


class B2C2CompareVersionAdmin(CompareVersionAdmin):
    def _reversion_revisionform_view(
        self, request: HttpRequest, *args: object, **kwargs: object
    ) -> HttpResponse:
        """
        Wrap this method from the original implementation to mute the django built-in signals.
        Otherwise, we may run into issues if signal handlers make changes to the outside world (e.g., publish to Redis)
        without using `transaction.on_commit` to ensure that the changes are made only after the database transaction is
        committed.
        For context on the issue, see the following section in `django-reversion` docs:
        - https://django-reversion.readthedocs.io/en/latest/admin.html#:~:text=your%20admin%20site.-,Important,-Using%20the%20admin
        The problematic line is this:
        - https://github.com/etianen/django-reversion/blob/fedcc716f272a227ae9cd02fab540dd8ec6516b8/reversion/admin.py#L172
        Note, that we still want to fire signals when the request method is POST, as this is the case when an object is
        restored from the history in the admin interface.
        """
        if request.method == "POST":
            context = nullcontext()
        else:
            context = factory.django.mute_signals(
                pre_save, post_save, pre_delete, post_delete, m2m_changed
            )
        with context:
            return super()._reversion_revisionform_view(request, *args, **kwargs)
``` 

The `factory.django.mute_signals` function comes from the [factory-boy](https://github.com/FactoryBoy/factory_boy/blob/68feb45e182f9acccfde671b7ba0babb5bc7ce11/factory/django.py#L275) project.

As you can see we simply prevent all Django built-in signals from being fired while a user is browsing history items through Django admin.

IMO it would make sense to include this change in `django-reversion` project. This could save others from unwanted and very nasty side-effects. Also at work we use this library in at least 3 projects so this would save us from having to implement the workaround in all 3+ projects...

What do you think? Would it be OK to reopen this ticket? I'd be happy to contribute the change if that helps.

**Comment by etianen** [comment]
I've just released this as v6.0.0.

I'm afraid my time for maintaining open source projects is very limited these days... but I try to get there eventually!
